sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/handler/DeliveryHandler.java

```diff
@@ -0,0 +1,489 @@
+            final ErrorCondition condition = link.getRemoteCondition();
+            addErrorCondition(logger.atVerbose(), condition)
+                .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                .addKeyValue(LINK_NAME_KEY, receiveLinkName)
+                .addKeyValue("updatedLinkCredit", link.getCredit())
+                .addKeyValue("remoteCredit", link.getRemoteCredit())
+                .addKeyValue("delivery.isPartial", delivery.isPartial())
+                .addKeyValue("delivery.isSettled", wasSettled)
+                .log("onDelivery.");
+        }
+    }
+
+    /**
+     * Completes the {@link Flux} of {@link Message} returned by {@link DeliveryHandler#getMessages()}.
+     *
+     * @param errorMessage message to log if completion fails.
+     */
+    void onDeliveryComplete(String errorMessage) {
+        messages.emitComplete((signalType, emitResult) -> {
+            logger.atVerbose()
+                .addKeyValue(ENTITY_PATH_KEY, entityPath)
+                .addKeyValue(LINK_NAME_KEY, receiveLinkName)
+                .addKeyValue(EMIT_RESULT_KEY, emitResult)
+                .log(errorMessage);
+            return false;
+        });
+    }
+
+    /**
+     * Gets the {@link Flux} that streams the {@link Message} objects decoded from the {@link Delivery}
+     * received from the broker.
+     *
+     * @return the {@link Flux} streaming {@link Message}.
+     */
+    Flux<Message> getMessages() {
+        return messages.asFlux();
+    }
+
+    /**
+     * Request settlement of delivery (with the unique {@code deliveryTag}) by sending a disposition frame
+     * with a state representing the desired-outcome, which the application wishes to occur at the broker.
+     * Disposition frame is sent via the same amqp receive-link that delivered the delivery, which was
+     * notified through {@link UnsettledDeliveries#onDelivery(UUID, Delivery)}.
+     *
+     * @param deliveryTag the unique delivery tag identifying the delivery.
+     * @param desiredState The state to include in the disposition frame indicating the desired-outcome
+     *                    that the application wish to occur at the broker.
+     * @return the {@link Mono} upon subscription starts the work by requesting ProtonJ library to send
+     * disposition frame to settle the delivery on the broker, and this Mono terminates once the broker
+     * acknowledges with disposition frame indicating outcome (a.ka. remote-outcome).
+     * The Mono can terminate if the configured timeout elapses or cannot initiate the request to ProtonJ
+     * library.
+     */
+    Mono<Void> sendDisposition(String deliveryTag, DeliveryState desiredState) {
+        return unsettledDeliveries.sendDisposition(deliveryTag, desiredState);
+    }
+
+    /**
+     * Prepare for the closure of this {@link DeliveryHandler}.
+     *
+     * @return a {@link Mono} that completes upon the completion of close preparation phase.
+     */
+    Mono<Void> prepareClose() {
```

Comment:
I think this method name is not clear enough; still thinking of a better name. So the idea was - this method is like to close() but co-exists with it, this attempts a "graceful closure" while close() performs a "forced closure". For example, when the user initiates receiver close, waiting for the graceful completion of pending tasks still makes sense since the link may be healthy (as we decide when to close the link). On the other hand, if the service initiates close, such a graceful closure may not be possible; we want immediately complete any pending work. In fact we're doing something similar in SBReactorReceiver, though not obvious. I added detailed java doc to this API in the last commit. 

Let me wire this graceful close and forced close in the upcoming commits so we can see if it makes sense.

Reply:
I've renamed the method and updated the JavaDoc to make it clear the purpose of this impl API.